### PR TITLE
KAFKA-14076: Fix issues with KafkaStreams.CloseOptions

### PR DESCRIPTION
**Pull request by jnh5y**
* Addresses issues with `KafkaStreams.close(CloseOptions)`.
* Adds an integration test for this new functionality.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java
Comment (jnh5y) [inline streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java:64]: I'm not wild about this IT as written.  I copied from the `AbstractResetIntegrationTest` and I'd be happy to hear a suggestion on how to make a more minimal test.

File: streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java]: nit: remove comment

File: streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java]: seems there is no `groupby` ?

File: streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java]: Because we actively remove members from the group, why do we need a timeout? Could we set timeout to zero?

In the end, a static member would be removed by the broker after `session.timeout.ms` passed -- we could also set a very high/infinit (`MAX_VALUE`) session timeout to ensure members are not removed by the broker but by the explicit "remove" command we send.

File: streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java:161]: Might be good to add a comment why we want to test with 2 threads (I understand why, but somebody else might miss context).

File: streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java]: Guess we could remove the map() step and just copy data from input to output? In the end, we just want to verify that we are running.

Comment (jnh5y) [reply streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java]: I tried `input.to(OUTPUT_TOPIC, ...)` but that broke the test.

Is this ok as-is?

Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java]: Did you upadate the serdes?
```
// String for the value instead of Long
.to(OUTPUT_TOPIC, Produced.with(Serdes.Long(), Serdes.String()));

// also for the `resultConsumerConfig` value L133 above:
resultConsumerConfig.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
```

File: streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java
Comment (jnh5y) [inline streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java]: The test takes at least 20 seconds to run.  This makes me wonder if the removal from the CG is working properly....

Comment (jnh5y) [reply streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java]: @mjsax until we figure this out, we shouldn't merge this PR.

File: streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java
Comment (jnh5y) [reply streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java:161]: Added a comment.

File: streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java
Comment (jnh5y) [reply streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java]: +1; done.

File: streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java
Comment (jnh5y) [reply streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java]: Removed the comment.

File: streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java
Comment (jnh5y) [reply streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java]: I've set it to MAX_VALUE.

Comment (jnh5y) [reply streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java]: Actually just changed my mind and have set it to 0.

File: streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java
Comment (jnh5y) [reply streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java]: Ok, I think I figured it out....

There was a little non-determinism in the `close()` call.  

The first approach I was using removed the consumers from the CG.  I *think* that if the StreamThreads were closed quickly enough, everything was fine.  If not, the StreamThreads re-subscribed to the CG and then we had to wait for the SESSION_TIMEOUT in order for the CG to be empty again.

I've updated the `close(CloseOptions)` method to close the StreamThreads first.  This means that `close` has to get a new Admin in order to remove things from the CG.

The only other option I could see would be pushing the work of leaving the CG to each StreamThread.  Maybe that's a better approach.  What do you think?

Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java]: Oh. Good, find. I thought/expected that the thread were close first, not afterwards...

But there is only one admin client that is shared over all threads, so why can't we keep admin open, send the "remove" request, and close admin afterwards (to avoid creating a new admin)? Guess not idea to "split" the logic and close admin outside of the centralized "private close()" method.

I am also fine if each thread removes only itself, after it called consumer.close(). Might even be more elegant if it's not too clumsy to implement.

Comment (jnh5y) [reply streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java]: Ok, I moved leaving the CG into the shutdown helper.  I think that'll do it...

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (jnh5y) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Of all the refactoring, this is probably the "scariest".  @mjsax thoughts?

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Do we still need this helper? Seems it's called only in a single place?

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Why can't we pass `timeoutMs`?

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:1530]: Seems we are missing a non-null check for `options`? (Unrelated to this PR, but we should just fix it on the side -- should be the first line of code in this method.)

File: streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java]: Should we set this to `MAX_VALUE` ?

File: streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java]: I think we should pass in a timeout via `CloseOptions` to overwrite default MAX_VALUE?

File: streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java]: What about this minor cleanup?

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (jnh5y) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Good catch.  Removing it.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (jnh5y) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: What `timeoutMs`?  

There's no `timeoutMs` in scope.

The method `closeToError` is called a few places; not sure if we could add a timeout to each of them...

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (jnh5y) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:1530]: +1.

File: streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java
Comment (jnh5y) [reply streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java]: That change is breaking things on my laptop.

The KafkaStreams application doesn't start in time...

File: streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java
Comment (jnh5y) [reply streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java]: Ok, we need to pick a value that will allow for states to transition and for a blocking call to leave the CG to return.  If we are low, the test may be flaky.

Does 1 second seem reasonable?  Or should it be higher?

File: streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java
Comment (jnh5y) [reply streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java]: I tried changing it to 
```        input.to(OUTPUT_TOPIC, Produced.with(Serdes.Long(), Serdes.String()));```
and the test hung on my laptop.

Am I doing something silly?

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Ah. My bad. I thought this is `close(timeout)`...

But if we `closeOnError` we don't send "remove" request and thus it should be fine as it's unused? Could we make it an `Optional` or `Long` (and pass `null`)?

File: streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java]: Where does it hang? Did you also change the _result_ consumer config so it can deserialize the output data on read?

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]:
```suggestion
        Objects.requireNonNull(options, "options cannot be null");
```

File: streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java]:
```suggestion
        streams.close(new CloseOptions().leaveGroup(true).timeout(Duration.ofMillis(30_000L)));
```

Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java]: We should give a little more head room for the "remove" request to finish. 1sec might make the test flaky on Jenkins.

File: streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java]:
```suggestion
        CLUSTER.deleteAllTopicsAndWait(120_000L);
```

File: streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java]: The error message is that we violate a broker max config setting. We should add
```
        brokerProps.setProperty(KafkaConfig.GroupMaxSessionTimeoutMsProp(), Integer.toString(Integer.MAX_VALUE));
```
inside the `static` block that creates `EmbeddedKafkaCluster`

This allows us to set `MAX_VALUE` here (btw: no need to call `toString()`

Comment (jnh5y) [reply streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java]: Nicely done.

File: streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java
Comment (jnh5y) [reply streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java]: Changing the result consumer config fixed it.

File: streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java:121]: Would be worth a comment to explain why this setting is important

File: streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java]:
```suggestion
        IntegrationTestUtils.startApplicationAndWaitUntilRunning(singletonList(streams), Duration.ofSeconds(30));
```

Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java]: Jenkins can be slow; 2 sec is very low and might make the test flaky. We usually use 30sec.

File: streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java]:
```suggestion
        streams.close(new CloseOptions().leaveGroup(true).timeout(Duration. ofSeconds(30)));
```

**Review by mjsax — APPROVED** [review]
Some final nits. Overall LGTM.

Thanks for pushing this through so we can ship this feature in 3.3.

File: streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java
Comment (jnh5y) [reply streams/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsCloseOptionsIntegrationTest.java:121]: Added a comment.

**Comment by mjsax** [comment]
Thanks for the fix. Merged to `trunk` and cherry-picked to `3.3` branch.

**Pull request merged**
